udf-compiler/src/test/scala/com/nvidia/spark/OpcodeSuite.scala

```diff
@@ -45,19 +42,1790 @@ class OpcodeSuite extends FunSuite {
 
   import spark.implicits._
 
-// Utility Function for checking equivalency of Dataset type
+  // Utility Function for checking equivalency of Dataset type
   def checkEquiv[T](ds1: Dataset[T], ds2: Dataset[T]) : Unit = {
-    ds1.explain(true)
-    ds2.explain(true)
+    assert(udfIsCompiled(ds1))
+    compareResult(ds1, ds2)
+  }
+
+  def checkEquivNotCompiled[T](ds1: Dataset[T], ds2: Dataset[T]): Unit = {
+    assert(!udfIsCompiled(ds1))
+    compareResult(ds1, ds2)
+  }
+
+  def compareResult[T](ds1: Dataset[T], ds2: Dataset[T]): Unit = {
     val resultdf = ds1.toDF()
     val refdf = ds2.toDF()
-    ds1.show
-    ds2.show
     val columns = refdf.schema.fields.map(_.name)
     val selectiveDifferences = columns.map(col => refdf.select(col).except(resultdf.select(col)))
     selectiveDifferences.map(diff => { assert(diff.count==0) } )
-    ds1.show
-    ds2.show
   }
-}
 
+  def udfIsCompiled[T](ds: Dataset[T]): Boolean = {
+    !ds.queryExecution.analyzed.toString().contains("UDF")
+  }
+
+  // START OF TESTS
+
+  test("not boolean") {
+    val myudf: Boolean => Boolean = { x => !x }
+    val u = makeUdf(myudf)
+    val dataset = List(true, false, true, false).toDS().repartition(1)
+    val result = dataset.withColumn("new", u('value))
+    val equiv = dataset.withColumn("new", not(col("value")))
+    checkEquiv(result, equiv)
+  }
+
+  test("and boolean") {
+    val myudf: (Boolean, Boolean) => Boolean = { (a, b) => a && b }
+    val u = makeUdf(myudf)
+    val df = List((false, false), (false, true), (true, false), (true, true))
+      .toDF("x", "y")
+      .repartition(1)
+    val result = df.withColumn("new", u(col("x"), col("y")))
+    val equiv = df.withColumn("new", col("x") && col("y"))
+    checkEquiv(result, equiv)
+  }
+
+  test("not(and boolean)") {
+    val myudf: (Boolean, Boolean) => Boolean = { (a, b) => !(a && b) }
+    val u = makeUdf(myudf)
+    val df = List((false, false), (false, true), (true, false), (true, true))
+      .toDF("x", "y")
+      .repartition(1)
+    val result = df.withColumn("new", u(col("x"), col("y")))
+    val equiv = df.withColumn("new", col("x") && col("y"))
```

Comment:
there's a bug in this test, and the unit tests are somehow passing

`equiv` should be `!(col("x") && col("y"))`.

@wjxiz1992 @seanprime7 Can you take a look at `checkEquiv` and see if you can reproduce?

Reply:
Also the tests are adding around 30% overhead to the CI team, anything we can do in `checkEquiv` to improve it would be great. I think timing the difference of having `checkEquiv` run a count per column,  vs a count for the whole table, would be a good place to start (currently it does `diff.count == 0` within a loop of the columns).

Reply:
@abellina 
Okay.  I will take a look today.
In the meantime, can you make all the issues resolved as such?

Reply:
@seanprime7 all look resolved to me, except for this one we are discussing now.

Reply:
Note we are looking to improve CI times in general by making tests parallel, so the perf thing may be go away. For now, still curious on whether we can improve `checkEquiv` for correctness and hopefully for performance.

Reply:
yeah.  I think the correctness has higher priority and that's what I am looking into right now.

Reply:
@abellina I have fixed the issue with d10d1a5.
With this change, the execution time of `mvn test -DwildcardSuites=com.nvidia.spark.OpcodeSuite` has also reduced from 3 minutes to 48 seconds on my local machine.

Reply:
I have further improved the execution time of `mvn test -DwildcardSuites=com.nvidia.spark.OpcodeSuite`.  It now finishes in 44 seconds on my local machine.

Reply:
Sounds great. Thanks @seanprime7